internal/util/util.go

```diff
@@ -291,3 +305,73 @@ func MergeDedupe(s1 []string, s2 []string) []string {
 	}
 	return s1
 }
+
+// Sample samples a protobuf message based on the sampling strategy.
+// The output is deterministic given the same strategy.
+func Sample(m proto.Message, strategy *SamplingStrategy) proto.Message {
+	pr := m.ProtoReflect()
+	pr.Range(func(fd protoreflect.FieldDescriptor, value protoreflect.Value) bool {
+		fieldName := fd.JSONName()
+
+		// Clear the excluded fields
+		for _, ex := range strategy.Exclude {
+			if ex == fieldName {
+				pr.Clear(fd)
+				return true
+			}
+		}
+
+		// If a field is not in the sampling strategy, keep it.
+		strat, ok := strategy.Children[fieldName]
+		if !ok {
+			return true
+		}
+		// Note, map[string]proto.Message is treated as protoreflect.MessageKind,
+		// So here need to check field and list first.
+		if fd.IsList() {
+			// Sampmle list.
+			oldList := value.List()
+			inc := int(1 / strat.Ratio)
+			var newList protoreflect.List
+			for i := 0; i < oldList.Len(); i += inc {
+				newList.Append(oldList.Get(i))
+			}
+			pr.Set(fd, protoreflect.ValueOfList(newList))
+		} else if fd.IsMap() {
```

Comment:
i wonder if we should skip sampling for maps. it'll be hard to keep the sample stable as the keys change.

Reply:
Also the time series are stored as map, so map sampling is needed.